src/main/java/com/cloudbees/jenkins/support/impl/HeapHisto.java

```diff
@@ -0,0 +1,104 @@
+
+import javax.management.InstanceNotFoundException;
+import javax.management.MBeanException;
+import javax.management.MalformedObjectNameException;
+import javax.management.ObjectName;
+import javax.management.ReflectionException;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.lang.management.ManagementFactory;
+import java.util.Collections;
+import java.util.Set;
+import java.util.WeakHashMap;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+@Extension
+public class HeapHisto extends Component {
+    private static final int OFFSET = 3;
+    private static final int MAX = 200 + OFFSET;
+
+    private static final Logger logger = Logger.getLogger(HeapHisto.class.getName());
+    private final WeakHashMap<Node, String> heapHistoCache = new WeakHashMap<Node, String>();
+
+    @NonNull
+    @Override
+    public Set<Permission> getRequiredPermissions() {
+        return Collections.singleton(Jenkins.ADMINISTER);
+    }
+
+    @NonNull
+    @Override
+    public String getDisplayName() {
+        return "Master Heap Histogram -histo:live";
+    }
+
+    @Override
+    public void addContents(@NonNull Container result) {
+        result.add(
+            new Content("nodes/master/heap-histogram.txt") {
+                @Override
+                public void writeTo(OutputStream os) throws IOException {
+                    os.write(getLiveHistogram(Jenkins.getInstance()).getBytes("UTF-8"));
+                }
+            }
+        );
+    }
+
+    private String getLiveHistogram(Node node) throws IOException {
+        return AsyncResultCache.get(node,
+                heapHistoCache,
+                new GetLiveHeapHistogram(),
+                "heap histogram",
+                "N/A: No connection to node, or no cache.");
+    }
+
+    private static final class GetLiveHeapHistogram implements Callable<String, RuntimeException> {
+        public String call() {
+            final String raw = getRawLiveHistogram();
+            final String[] lines = raw.split("\n");
+            final int limit = MAX <= lines.length ? MAX : lines.length;
+
+            final StringBuilder bos = new StringBuilder();
+
+            bos.append("Master Heap Histogram -histo:live");
+            for (int i=0; i<limit; i++) {
+                bos.append(lines[i]).append('\n');
+            }
+
+            return bos.toString();
+        }
+
+        /** {@inheritDoc} */
+        @Override
+        public void checkRoles(RoleChecker checker) throws SecurityException {
+            // TODO: do we have to verify some role?
+        }
+
+        private String getRawLiveHistogram() {
+            String result;
+            try {
+                ObjectName objName = new ObjectName("com.sun.management:type=DiagnosticCommand");
+                result = (String) ManagementFactory.getPlatformMBeanServer().invoke(objName, "gcClassHistogram", new Object[] {null}, new String[]{String[].class.getName()});
```

Comment:
I'd guess null result is not possible here. Other implementations like [nuxeo-connect-tools](https://github.com/pipeline-as-code-nxmatic/nuxeo-connect-tools/blob/4423de6c0cfa7435d9f198b1b123a2ddc324ed6a/report-core/src/main/java/org/nuxeo/connect/tools/report/management/MxClassHistogramReport.java) do not check it as well